### PR TITLE
[ROCm] Encoder prefill attention on gfx115x: split-D head dims, alignment hint, launch config

### PR DESCRIPTION
**Pull request by roberteg16**
## Summary

`vllm/v1/attention/ops/triton_prefill_attention.py` is the **encoder** attention kernel — ViT towers via the mm-encoder backend, the encoder branches of `TRITON_ATTN` and `ROCM_ATTN`, and `mimo_v2_omni`'s visual attention. (The paged LLM prefill path imports a same-named `context_attention_fwd` from `prefix_prefill.py`; that is a different module and is untouched here.)

It has no launch configuration of its own, and it pads the head-dim tile to `next_power_of_2(head_dim)`, which wastes most of the matmul on the non-power-of-2 head dims that vision towers use. On gfx1151 this series takes Gemma3-4B's TTFT down **41%** and Qwen2.5-VL-7B's down **15%**.

Everything is gated on gfx115x, the only architecture measured; every other architecture keeps today's behaviour exactly. The split and the alignment hint are mechanically vendor-neutral and should carry to CDNA and NVIDIA, but I have no hardware to confirm that, and an unmeasured regression on someone else's GPU is worse than a missed win. Widening the gate later is a branch in `_get_head_dim_blocks` / `_get_encoder_launch`.

## Commits

| commit | what it does |
|---|---|
| `[ROCm] Add a gfx115x architecture predicate` | `_ON_GFX115X` / `on_gfx115x()`. The other two ROCm commits gate on it, so it lands first. |
| `[ROCm] Cover non-power-of-2 head dims with a split-D block on gfx115x` | `head_dim=72` ran the `qk`/`pv` dots over 128 lanes when only 5 K-passes of 16 are needed. Covers the head dim as a power-of-2 main block plus a power-of-2 tail block (72 → 64+16 = 80) behind a new constexpr `BLOCK_DMODEL_TAIL`; a tail of 0 is dead code and reproduces the single-block kernel exactly. `_split_head_dim` only takes the tail when it strictly narrows the extent — 112 would split into 64+64, covering the same 128 lanes for 12% more time. |
| `[Attention] Hint 16-byte head-stride alignment in ViT prefill attention` | When `head_dim % 16 == 8`, Triton does not attach `tt.divisibility=8` to the head stride, so the D-contiguous Q/K/V loads lower to scalar 2-byte loads. Hints `tl.multiple_of(off_h_*, 8)` behind a constexpr the wrapper sets from the **actual runtime strides**, so it stays sound for non-contiguous views. |
| `[ROCm] Give the encoder prefill attention a gfx115x launch config` | `BLOCK_N` inherits `BLOCK_M`, so gfx115x gets a 128x128 square tile chosen for NVIDIA. `_get_encoder_launch` dispatches on architecture and `_gfx115x_encoder_launch` holds the tuning, so another architecture is a branch in the former rather than a change to the latter. |

## Performance

Radeon 8060S (gfx1151). Each cell carries that commit's own contribution in brackets; the predicate commit is a no-op and is omitted.

**Kernel** — `context_attention_fwd`, B=1, S=3520, H=16, bf16, non-causal, ms/call. Each pass checks out every point in turn and the table is the median of 3 such passes, so drift cannot be mistaken for a commit's effect:

| head_dim | baseline | `+split-D` | `+alignment hint` | `+launch config` | total |
|---:|---:|---:|---:|---:|---:|
| 64 | 4.519 | 4.530 (1.00x) | 4.522 (1.00x) | 1.856 (**2.44x**) | **2.43x** |
| 72 | 9.265 | 7.536 (**1.23x**) | 6.946 (**1.08x**) | 2.233 (**3.11x**) | **4.15x** |
| 80 | 8.674 | 7.423 (**1.17x**) | 7.440 (1.00x) | 2.351 (**3.16x**) | **3.69x** |
| 128 | 16.331 | 16.087 (1.02x) | 16.082 (1.00x) | 3.907 (**4.12x**) | **4.18x** |
| 256 | 29.949 | 29.943 (1.00x) | 29.985 (1.00x) | 13.711 (**2.19x**) | **2.18x** |

split-D moves only the non-power-of-2 head dims, as it must — elsewhere the tail is dead code. The alignment hint moves only `head_dim=72`, the only dim measured with `head_dim % 16 == 8`.

**End to end** — TTFT, single image, `max_tokens=1`, prefix caching and the multimodal processor cache disabled and a different image per repetition so every call really re-runs the vision tower. Median of 5, ms:

| model | ViT head_dim | image | baseline | `+split-D` | `+alignment hint` | `+launch config` | total |
|---|---:|---|---:|---:|---:|---:|---:|
| `gemma-3-4b-it-quantized.w4a16` (SigLIP) | 72 | 1920x1080 | 604.6 | 545.8 (**−9.7%**) | 523.4 (**−4.1%**) | 356.6 (**−31.9%**) | **−41.0%** |
| `Qwen2.5-VL-7B-Instruct` | 80 | 1920x1080 | 2296.3 | 2203.3 (**−4.0%**) | 2219.8 (+0.7%) | 1947.8 (**−12.3%**) | **−15.2%** |
| `Qwen2.5-VL-7B-Instruct` | 80 | 1024x800 | 786.8 | 762.7 (**−3.1%**) | 766.3 (+0.5%) | 717.1 (**−6.4%**) | **−8.9%** |

Gemma3 pads to a fixed 896x896 and pairs a 4B w4a16 language model with a 27-layer tower, so the vision encoder dominates its TTFT; Qwen2.5-VL-7B spends proportionally more time in the language model. Both runs were checked to really reach this kernel — 27 calls per request at shape `(16, 72)` and 32 at `(16, 80)`, matching each tower's depth.

## Why these launch-config values

A sweep of `BLOCK_M` x `BLOCK_N` x `num_warps` x `waves_per_eu` puts the optimum at `BLOCK_N=16` with 4 warps up to a head dim of 80 and `BLOCK_N=32` with 8 warps above it; at 64–80 the wider pairing is 20–35% slower and from 96 up the ordering reverses. No `waves_per_eu` override helped anywhere, so none is set, and `BLOCK_M` keeps the generic selection.

This is not overfitted to the two ViT shapes that motivated it: against the generic tile, **18 of 18 `head_dim` x `seqlen` cells improve, 2.04x to 4.02x** (head dims 64–256, seqlens 1024–8192), with no regression anywhere. Envelope checks at S=2048 also all improve — sliding window 512 (1.85x–3.35x), GQA with 16 query heads over 4 KV heads (2.84x–3.09x), fp16 (2.53x–3.41x).

`benchmarks/kernels/benchmark_vit_prefill_attn.py` is added here so any of this is reproducible; it launches `_fwd_kernel` directly so `--bm/--bn/--nw/--ns/--we` are honoured.

```
python benchmarks/kernels/benchmark_vit_prefill_attn.py --seq 3520 --heads 16 --head-dim 72 \
    --backend TRITON_ATTN --bm 128 --bn 16 --nw 4
```

## Testing

```
pytest tests/kernels/attention/test_triton_prefill_attention.py   125 passed
pre-commit run --files <changed files>                            clean (ruff, ruff-format, mypy, SPDX)
```

The existing correctness tests only covered `head_dim` 128; they now also run 64, 72, 80 and 96, which between them exercise both tail widths the kernel can emit (72 and 80 split as 64+16, 96 as 64+32). Because the split is gated, a further test forces it on so the path keeps its coverage on CI machines that are not gfx115x, and `_split_head_dim` is checked exhaustively over head dims 1 to 512 in pure Python. Other new tests cover a view whose head stride is *not* 8-element aligned, and the launch-config selection on both sides of the head-dim threshold.

Each commit compiles and passes its own tests in isolation (20, 111, 113 and 125 passed), so the series stays bisectable. Through the first two commits every power-of-2 head dim is bit-identical in bf16 and fp32; the launch config then changes the KV tile, which changes the order the softmax accumulates in, so bf16 output moves by up to one ULP (max absolute deviation 7.8e-03 at `head_dim` 64, exactly `torch.finfo(bfloat16).eps`).

## Not a duplicate

No open PR touches `triton_prefill_attention.py`. #56232, #56253 and #56276 tune `triton_unified_attention.py` (decoder prefill/decode) — the same argument that gfx11 needs its own launch config, applied to the other attention kernel. They do not conflict: encoder attention reaches this kernel under both `ROCM_ATTN` and `TRITON_ATTN`, so #56232's backend reordering does not change whether this tune applies.

**One known overlap:** #56307 adds the same `_ON_GFX115X` / `on_gfx115x()` predicate to `vllm/platforms/rocm.py`. Whichever lands first, the other drops that commit and rebases.

## AI assistance

This change was prepared with AI assistance (Claude). Every line was reviewed by the submitter, and the measurements above were run and checked by hand.

### CODE REVIEW

**Comment by roberteg16** [comment]
Upstream PR: https://github.com/vllm-project/vllm/pull/56450

**Comment by roberteg16** [comment]
Superseded by #1322. This PR was rebased onto current upstream, which now includes #58225; GitHub would not reattach the force-pushed branch to this PR, so it was reopened as #1322 with the same branch and base.
